resolwe/flow/models/utils/dynamic_resources.py

```diff
@@ -0,0 +1,103 @@
+"""Utility functions for dynamic calculation of process resource requirements."""
+
+import ast
+import dis
+import math
+from functools import reduce
+
+from resolwe.flow.models import Data, Process
+
+
+def _eval_dynamic_requirements(expr: str, variables: dict):
+    """Evaluate an arithmetic expression."""
+    allowed_codes = {
+        "RESUME",  # noop
+        "BINARY_OP",  # infix binary op
+        "BINARY_ADD",  # +
+        "BINARY_SUBTRACT",  # -
+        "BINARY_MULTIPLY",  # *
+        "BINARY_TRUE_DIVIDE",  # /
+        "BINARY_FLOOR_DIVIDE",  # //
+        "BINARY_MODULO",  # %
+        "BINARY_POWER",  # **
+        "COMPARE_OP",  # == > < >= <=
+        "UNARY_NOT",  # not
+        "LOAD_CONST",  # constants
+        "LOAD_NAME",  # variables
+        "CALL_FUNCTION",  # provided functions, e.g. log
+        "RETURN_VALUE",  # implicit return
+        "RETURN_CONST",  # implicit return (alt.)
+    }
+    try:
+        compiled = compile(expr, "", "eval")
+    except SyntaxError:
+        raise ValueError(f"{expr} is not a valid expression")
+
+    for instruction in dis.get_instructions(expr):
+        opcode = dis.opname[instruction.opcode]
+        if opcode not in allowed_codes:
+            raise ValueError(f"Operation {opcode} not allowed (in {expr})")
+
+    scope = {"log": math.log, "exp": math.exp, "sqrt": math.sqrt}
+    scope.update(variables)
+    scope.update({"__builtins__": {}})
+    return eval(compiled, scope)
+
+
+def schema_get(schema: list, key: str) -> dict | list:
+    """Return a schema field with the given name."""
+    field = next(iter(filter(lambda field: field["name"] == key, schema)))
+    if "group" in field:
+        return field["group"]
+    return field
+
+
+def _resolve_value(name: str, input_values: dict, input_schema: list):
+    """Resolve the value of a variable in the expression."""
+
+    try:
+        keys = name.split("__")
+        value = reduce(dict.get, keys, input_values)
+        type_ = reduce(schema_get, keys, input_schema)["type"]
+    except (KeyError, StopIteration):
+        raise ValueError(f"Unrecognized variable {name}")
+
+    # More types may be supported as needed.
+    if type_ == "data:":
+        return Data.objects.get(pk=value).size
+    elif type_ in ("basic:integer:", "basic:float:", "basic:boolean:"):
+        return value
+
+    raise ValueError(f"Unsupported type {type_}")
+
+
+def get_dynamic_resource_limits(process: Process, data: Data):
```

Comment:
Not sure if this belongs here or in the process class. 

I do not see any added value by having these methods in the `utils` file, since they will probably only be used in the `process.py`  model file.

Reply:
While it's true the functions won't be reused, I feel keeping the expression evaluation logic separate simplifies maintenance by isolating complex calculation logic from the core ORM functionalities of the `Process` class, as they are only marginally related.